### ISSUE TITLE
Web interface not supporting color calibration via Kodi

### ISSUE THREAD

**Issue by Nears** [hyperion-project/hyperion.ng#771]
- [x] I confirm that this is an issue rather than a question.

## Bug report
It is not possible to run the color calibration via web-interface. Even if Kodi is running and the right IP address is set the calibration does not find Kodi.

#### Steps to reproduce
Install Kodi (Version 18.6.) and Hyperion.NG (Version 2.0.0.Alpha 4) (From https://github.com/hyperion-project/hyperion.ng/releases/download/2.0.0-alpha.4/Hyperion-2.0.0-alpha.4-Linux-armv7hf-rpi.deb)

#### What is expected?
Calibration mode is started and colors can ge calibrated

#### What is actually happening?
Calibration does not start and times out after some time. User is requested to start manual calibration, but even during this the Color-Picker GUI shows up but the LEDs are not changing theire color. Therefor no calibration is possible.

#### System
Hyperion Server: 
- Build:       (HEAD detached at 731b517b) (brindosch-80f7da4d/731b517b-1585347672)
- Build time:  Mar 28 2020 12:26:15
- Git Remote:  https://github.com/hyperion-project/hyperion.ng
- Version:     2.0.0-alpha.4
- UI Lang:     auto (BrowserLang: de)
- UI Access:   default
- Avail Capt:  dispmanx,v4l2,framebuffer,qt

Hyperion Server OS: 
- Distribution: Raspbian GNU/Linux 10 (buster)
- Arch:         arm
- Kernel:       linux (4.19.97-v7l+ (WS: 32))
- Browser:      Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:75.0) Gecko/20100101 Firefox/75.0

**Comment by pattyland** [comment]
Did you also activate the web remote in Kodi? I use Alpha 5 and MrMC and the calibration via web interface worked fine

**Comment by FransTwisk** [comment]
I'm experiencing the same. Pre-filled IP is correct but I can't click "Continue" because it's disabled. Hyperion & Kodi 18.6 running on the same Raspberry Pi 4 running Raspbian Buster. It also didn't work on my LibreELEC install.

If I change the IP and port to "localhost:8080" / "127.0.0.1:8080" and press enter, it says "Kodi webserver not found, proceed without Kodi support". Kodi webserver is running and works if I visit it myself. This is what the call being made:

```
GET http://localhost:8080/jsonrpc?callback=jsonCallback&request=%7B%22jsonrpc%22%3A%222.0%22%2C%22method%22%3A%22GUI.ShowNotification%22%2C%22params%22%3A%7B%22title%22%3A%20%22Colour%20calibration%20wizard%22%2C%20%22message%22%3A%20%22Test%20success%20-%20time%20to%20proceed!%22%2C%20%22image%22%3A%22info%22%2C%20%22displaytime%22%3A5000%20%7D%2C%22id%22%3A%221%22%7D&_=1588410136436 net::ERR_CONNECTION_REFUSED
```

Under Query String Parameters I see:
```
callback: jsonCallback
request: {"jsonrpc":"2.0","method":"GUI.ShowNotification","params":{"title": "Colour calibration wizard", "message": "Test success - time to proceed!", "image":"info", "displaytime":5000 },"id":"1"}
_: 1588410326952
```

**Comment by Nears** [comment]
> 
> 
> Did you also activate the web remote in Kodi? I use Alpha 5 and MrMC and the calibration via web interface worked fine

Yes its active in Kodi and working, I also use it for remote controll access.

**Comment by LichtiMC** [comment]
In current release (2.0.0-alpha.7) I have the same problem.
How would I calibrate the colors now?

Thanks...

**Comment by m-seker** [comment]
> I'm experiencing the same. Pre-filled IP is correct but I can't click "Continue" because it's disabled. Hyperion Alpha 5 & Kodi 18.6 running on the same Raspberry Pi 4 running Raspbian Buster. It also didn't work on my LibreELEC install. This first call actually works, with the prefilled local IP, but returns:
> 
> ```
> jsonCallback({"error":{"code":-32099,"message":"Bad client permission."},"id":"1","jsonrpc":"2.0"});
> ```
> 
> This is what shows up in the Kodi debug log:
> 
> ```
> 2020-05-02 11:18:20.089 T:2052059344   DEBUG: CWebServer[8080]: request received for /jsonrpc?callback=jsonCallback&request={"jsonrpc":"2.0","method":"GUI.ShowNotification","params":{"title": "Colour calibration wizard", "message": "Test success - time to proceed!", "image":"info", "displaytime":5000 },"id":"1"}&_=1588411096647
> ```
> 
> Allow remote control from applications on this system & Allow remote control from applications on other systems are turned on. Username and password are blank. Web remote works fine.

1- Did you figure out a way to overcome this issue ? Or did it just disappear with later releases(alpha6 or alpha7) ?

2- Are Raspbian and Kodi working fine on Pi4 ? I tried once and it was painfully slow, then I didn't check any further.

**Comment by FransTwisk** [comment]
> > I'm experiencing the same. Pre-filled IP is correct but I can't click "Continue" because it's disabled. Hyperion Alpha 5 & Kodi 18.6 running on the same Raspberry Pi 4 running Raspbian Buster. It also didn't work on my LibreELEC install. This first call actually works, with the prefilled local IP, but returns:
> > ```
> > jsonCallback({"error":{"code":-32099,"message":"Bad client permission."},"id":"1","jsonrpc":"2.0"});
> > ```
> > 
> > 
> > This is what shows up in the Kodi debug log:
> > ```
> > 2020-05-02 11:18:20.089 T:2052059344   DEBUG: CWebServer[8080]: request received for /jsonrpc?callback=jsonCallback&request={"jsonrpc":"2.0","method":"GUI.ShowNotification","params":{"title": "Colour calibration wizard", "message": "Test success - time to proceed!", "image":"info", "displaytime":5000 },"id":"1"}&_=1588411096647
> > ```
> > 
> > 
> > Allow remote control from applications on this system & Allow remote control from applications on other systems are turned on. Username and password are blank. Web remote works fine.
> 
> 1- Did you figure out a way to overcome this issue ? Or did it just disappear with later releases(alpha6 or alpha7) ?
> 
> 2- Are Raspbian and Kodi working fine on Pi4 ? I tried once and it was painfully slow, then I didn't check any further.

I did not. I did however switch to an Nvidia Shield, since I really missed not being able to play 4k & HDR content. Kodi works fine on a Pi 4, but no HDR support and now Hyperion support for 4k. I do however have the same issue on Nvidia Shield where Hyperion also can't connect to calibrate through Kodi.

**Comment by sPOiDar** [comment]
Since Kodi Leia non-read-only JSONRPC commands must be sent via POST rather than GET.

**Comment by Lord-Grey** [comment]
@sPOiDar Correct. Using Kodi's REST API via POST is causing all the CORS problems and does not work by using JS in the browser, In addition, now "Content-type" = "application/json" is mandatory, which makes the call a non-simple one and the CORS rules apply.

An alternative solution would have been, implementing KODI's REST API in the backend, which I avoided for the time being...

**Comment by gismo2004** [comment]
Websokets will work and i made already something similar some time ago... https://github.com/gismo2004/hyperion.ng/commit/da2e0f387ac9c7e8c4e4a1bb65ad0ee4f393a76e

**Comment by Lord-Grey** [comment]
@gismo2004 Maybe next time you share improvements via an PR that everybody could benefit. :)
Would have saved me some time, too....

**Comment by gismo2004** [comment]
Well, I have done that already, but obviously the deal in following PR was not fulfilled 😉
https://github.com/hyperion-project/hyperion.ng/pull/492

**Comment by Lord-Grey** [comment]
@gismo2004 😀

**Comment by rradar** [comment]
> Pre-filled IP is correct but I can't click "Continue" because it's disabled

Exactly the problem I have now in the year 2021.

- 2.0.0-alpha.8
- CoreElec
- Kodi 18.9
- Platform: | Amlogic
- Controller type: | wled

Tried with different browsers but no luck...

**Comment by Paulchen-Panther** [comment]
> > Pre-filled IP is correct but I can't click "Continue" because it's disabled
> 
> Exactly the problem I have now in the year 2021.
> 
> * 2.0.0-alpha.8
> * CoreElec
> * Kodi 18.9
> * Platform: | Amlogic
> * Controller type: | wled
> 
> Tried with different browsers but no luck...

2.0.0-alpha.8 is not up to date and CoreElec is not supported. Sorry

**Comment by rradar** [comment]
> 2.0.0-alpha.8 is not up to date

You are right! And most importantly: The fix for this bug (https://github.com/hyperion-project/hyperion.ng/pull/1044) was merged in the latest release 2.0.0-alpha.9

> CoreElec is not supported.

But works perfectly fine :yum:
